qiskit/circuit/library/arithmetic/inverse_chebyshev.py

```diff
@@ -0,0 +1,194 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2019, 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Piecewise polynomial approximation to arcsin(1/x)."""
+
+from typing import Optional
+import numpy as np
+from numpy.polynomial.chebyshev import Chebyshev
+
+from qiskit.circuit import QuantumRegister
+from qiskit.circuit.library.blueprintcircuit import BlueprintCircuit
+from qiskit.circuit.exceptions import CircuitError
+
+from .piecewise_polynomial_pauli_rotations import PiecewisePolynomialPauliRotations
+from .integer_comparator import IntegerComparator
+
+# pylint: disable=invalid-name
+
+
+class InverseChebyshev(BlueprintCircuit):
+    r"""Piecewise Chebyshev approximation of the inverse function.
+
+    Building block of HHL. For a given constant :math:`C` and error tolerance :math:`\epsilon`,
+    this class implements a piecewise polynomial approximation on :math:`n` qubits to
+    :math:`arcsin(C/x)`, where :math:`x \in [a,2^n - 1]` and :math:`a` depends on :math:`\epsilon`.
+
+    The values of the parameters are calculated according to https://arxiv.org/abs/2009.04484.
+    """
+
+    def __init__(self,
+                 num_state_qubits: Optional[int] = None,
+                 epsilon: Optional[float] = None,
+                 constant: Optional[float] = None,
+                 kappa: Optional[float] = None,
+                 name: str = 'inv_cheb') -> None:
+        """Construct the piecewise Chebyshev approximation of the inverse function.
+
+        Args:
+            num_state_qubits: number of qubits representing the state.
+            epsilon: accuracy of the approximation.
+                Defaults to ``1e-2``.
+            constant: :math:`C` in :math:`arcsin(C/x)`.
+                Defaults to ``1``.
+            kappa: condition number of the system.
+                Defaults to ``1``.
+            name: The name of the circuit object.
+        """
+        super().__init__(name=name)
+
+        # define internal parameters
+        self._num_state_qubits = None
+
+        # Store parameters
+        self._epsilon = epsilon if epsilon is not None else 1e-2
+        self._kappa = kappa if kappa is not None else 1
+        self._constant = constant if constant is not None else 1
+
+        # We perform the identity operation on [1,a].
+        # int(round()) necessary to compensate for computer precision.
+        self._N_l = 2 ** num_state_qubits
```

Comment:
All these computations should happen inside `_build`. In a `BlueprintCircuit`, users can change the `num_state_qubits` after the circuit is constructed. Then these computations must be done again --> into `_build` 🙂